src/main/kotlin/com/sourcegraph/cody/history/HistoryService.kt

```diff
@@ -0,0 +1,55 @@
+package com.sourcegraph.cody.history
+
+import com.intellij.openapi.components.*
+import com.sourcegraph.cody.agent.protocol.ChatMessage
+import com.sourcegraph.cody.agent.protocol.Speaker
+import com.sourcegraph.cody.history.state.ChatState
+import com.sourcegraph.cody.history.state.HistoryState
+import com.sourcegraph.cody.history.state.MessageState
+import java.time.LocalDateTime
+
+@State(name = "ChatHistory", storages = [Storage("cody_history.xml")])
+@Service(Service.Level.PROJECT)
+class HistoryService : SimplePersistentStateComponent<HistoryState>(HistoryState()) {
+
+  private val listeners = mutableListOf<(ChatState) -> Unit>()
+
+  fun listenOnUpdate(listener: (ChatState) -> Unit) {
+    synchronized(listeners) { listeners += listener }
+  }
+
+  fun update(internalId: String, chatMessages: List<ChatMessage>) {
+    val found = getChatOrCreate(internalId)
+    found.messages = chatMessages.map(::convertToMessageState).toMutableList()
+    found.setUpdatedTimeAt(LocalDateTime.now())
+    listeners.forEach { it(found) }
```

Comment:
You synchronised it in one place, but you need to do it in all of them.

Reply:
Good point! Fixed.